cdap-common/src/main/java/co/cask/cdap/common/security/AuthEnforceUtil.java

```diff
@@ -37,17 +54,113 @@ private AuthEnforceUtil() {
+    } else {
+      return createEntityId(entityClass, entities);
+    }
+  }
+
+  /**
+   * Return the required size of entity parts to create the {@link EntityId} on which authorization enforcement
+   * needs to be done as specified in {@link AuthEnforce#enforceOn()}
+   *
+   * @param enforceOn the {@link Type} of {@link EntityId} on which enforcement needs to be done
+   * @return the size of entity parts needed to create the above {@link EntityId}
+   * @throws IllegalArgumentException of the given enforceOn is not of supported {@link EntityId} type
+   */
+  static int getEntityIdPartsCount(Type enforceOn) {
+    if (enforceOn.equals(Type.getType(InstanceId.class))) {
+      return CONS_CACHE.get(InstanceId.class).getParameterTypes().length;
+    } else if (enforceOn.equals(Type.getType(NamespaceId.class))) {
+      return CONS_CACHE.get(NamespaceId.class).getParameterTypes().length;
+    } else if (enforceOn.equals(Type.getType(StreamId.class))) {
+      return CONS_CACHE.get(StreamId.class).getParameterTypes().length;
+    } else if (enforceOn.equals(Type.getType(DatasetId.class))) {
+      return CONS_CACHE.get(DatasetId.class).getParameterTypes().length;
+    } else if (enforceOn.equals(Type.getType(ApplicationId.class))) {
+      return CONS_CACHE.get(ApplicationId.class).getParameterTypes().length;
+    } else if (enforceOn.equals(Type.getType(ArtifactId.class))) {
+      return CONS_CACHE.get(ArtifactId.class).getParameterTypes().length;
+    } else if (enforceOn.equals(Type.getType(ProgramId.class))) {
+      return CONS_CACHE.get(ProgramId.class).getParameterTypes().length;
+    } else {
+      throw new IllegalArgumentException(String.format("Failed to determine required number of entity parts " +
+                                                         "needed for %s. Please make sure its a valid %s class " +
+                                                         "for authorization enforcement",
+                                                       enforceOn.getClassName(), EntityId.class.getSimpleName()));
+    }
+  }
+
+  private static EntityId createEntityId(Class<? extends EntityId> entityClass, Object[] args)
+    throws IllegalAccessException, InvocationTargetException, InstantiationException {
+    Constructor<? extends EntityId> constructor = CONS_CACHE.get(entityClass);
+
+    Preconditions.checkNotNull(constructor, String.format("Failed to find constructor for entity class %s. Please " +
+                                                            "make sure it exists.", entityClass));
+    // its okay to call with object [] without checking that all of these are string because if one of them is not
+    // then newInstance call will throw IllegalArgumentException.
+    return constructor.newInstance(args);
+  }
+
+  private static Constructor<? extends EntityId> findConstructor(Class<? extends EntityId> entityClass) {
+    Constructor<? extends EntityId> constructor = null;
```

Comment:
Simpler to write it with Java 8 lambda:

```java
    // Find the constructor with all String parameters
    for (Constructor<?> curConstructor : entityClass.getConstructors()) {
      if (Arrays.stream(curConstructor.getParameterTypes()).allMatch(String.class::equals)) {
        return (Constructor<? extends EntityId>) curConstructor;
      }
    }
    // since constructor was not found throw an exception
    throw new IllegalStateException(String.format("Failed to find constructor for %s whose parameters are only of " +
                                                  "String type", entityClass.getName()));
  }
```